meetings/working-groups/extensions/extending-extensions-a-guide-to-relaxation.md

```diff
@@ -0,0 +1,316 @@
+`param` attribute specifier just as they are in the base design.
+
+Unfortunately, adding an optional parameter name at the top of the extension declaration means there’s another axis
+that might force an extension author to need another declaration. If the author wants the `this` parameter name to
+change between extension members within the same underlying type, they’ll need a declaration for each variation of the
+parameter name.
+
+This can be solved with another relaxation.
+
+### Mixing underlying types within an extension declaration
+
+While the anonymous extension declaration proposal protects the consumer from the "type explosion" problem, that
+problem is still very much alive for the extension author. The base proposal makes the following assumptions:
+
+> - Underlying types belong together with their type parameter declarations, and it would be confusing to separate them.
+> - People will resent the verbosity of having to repeat underlying types and accompanying type parameters for each member.
+
+These assumptions drive a design in which all members within an extension declaration must share the same underlying
+type for their `this` parameter, including nullable reference types. While it is possible that both of these
+assumptions may true, they will not be true for all.
+
+It _may_ be confusing to separate underlying types from the their type parameter declarations. However, that depends
+largely on the mental model a programmer develops for extensions. If they see extensions as a special form of type
+inheritance, it would indeed by confusing to separate type parameters from the type itself! However, if they view
+extensions as a new way to extension methods that allow for other members with a simpler syntax, they might not be so
+confused.
+
+It’s also true that people may resent having to repeat underlying types and accompanying type parameters for each
+member. However, if the first assumption above is relaxed to allow the underlying type to be separated from their type
+parameters, the second assumption becomes much more palatable. And, classic extension methods require that the
+underlying type be repeated for each method, and it is a massively popular feature.
+
+Consider the following classic extension methods from Roslyn’s public API:
+
+```c#
+public static class CSharpExtensions
+{
+    public static bool IsKind(this SyntaxToken token, SyntaxKind kind) => ...;
+    public static bool IsKind(this SyntaxTrivia trivia, SyntaxKind kind) => ...;
+    public static bool IsKind([NotNullWhen(true)] this SyntaxNode? node, SyntaxKind kind) => ...;
+    public static bool IsKind(this SyntaxNodeOrToken nodeOrToken, SyntaxKind kind) => ...;
+    public static bool ContainsDirective(this SyntaxNode node, SyntaxKind kind) => ...;
+}
+```
+
+If these were written using the base design, they would need to be declared across four extension declarations:
+
+```c#
```

Comment:
If the braces are optional this doesn't feel too bad?

```cs
public static class CSharpExtensions
{
    extension(SyntaxToken) public bool IsKind(SyntaxKind kind) => ...;
    extension(SyntaxTrivia) public bool IsKind(SyntaxKind kind) => ...;

    extension(SyntaxNode?)
        [param: NotNullWhen(true)]
        public bool IsKind(SyntaxKind kind) => ...;

    extension(SyntaxNodeOrToken) public bool IsKind(SyntaxKind kind) => ...;
    extension(SyntaxNode) public bool ContainsDirective(SyntaxKind kind) => ...;
}
```

Reply:
> If the braces are optional this doesn't feel too bad?
> 
> ```cs
> public static class CSharpExtensions
> {
>     extension(SyntaxToken) public bool IsKind(SyntaxKind kind) => ...;
>     extension(SyntaxTrivia) public bool IsKind(SyntaxKind kind) => ...;
>     extension([NotNullWhen(true)] SyntaxNode?) public bool IsKind(SyntaxKind kind) => ...;
>     extension(SyntaxNodeOrToken) public bool IsKind(SyntaxKind kind) => ...;
>     extension(SyntaxNode) public bool ContainsDirective(SyntaxKind kind) => ...;
> }
> ```

This is definitely a possibility! It's a bit verbose for my personal taste though, and I'm not sure how I feel about it being ahead of all modifiers and the return type. I aimed for a syntax that corresponds with what an extension member looks like when it's called in its reduced form. The way I was thinking about it was to take the `this` parameter from a classic extension method and shift it to the left. Essentially, it embraces the idea that the "underlying type" is really just a parameter.

![image](https://github.com/user-attachments/assets/2b5f4d33-7d33-4a3c-8042-10dc21b422af)

Reply:
> tangentially related: since this is literally a sort of modifier on members (either inline or as a block), does this practically exclude interfaces (in which case youd have use another type kind) or they're going to appear together with nested extensions?

I have no idea. :smile: I'm expecting that a lot of design will have happened and moved things around before we get to extension interfaces.